### PR TITLE
[incubator/zookeeper] Normalize names, labels and selectors

### PR DESCRIPTION
**Pull request by h0tbird**
If your helm release name is `kafka` you get:

**Stateful Sets**
*kafka-zookeeper*

**Pods**
*kafka-zookeeper-0*
*kafka-zookeeper-1*
*kafka-zookeeper-2*

**Services**
*kafka-zookeeper* (clients)
*kafka-zookeeper-headless*

### CODE REVIEW

File: incubator/zookeeper/templates/PodDisruptionBudget.yaml [hunk 1]
```diff
@@ -1,20 +1,20 @@
...
-  name: "{{ printf "zk-pdb-%s" .Release.Name | trunc 24 }}"
+  name: {{ template "fullname" . }}
   annotations:
     helm.sh/created: {{.Release.Time.Seconds | quote }}
   labels:
     heritage: {{.Release.Service | quote }}
     release: {{.Release.Name | quote }}
     chart: "{{.Chart.Name}}-{{.Chart.Version}}"
-    component: "{{ printf "zk-%s" .Release.Name  | trunc 24 }}"
+    component: {{ template "name" . }}
```
Comment (unguiculus) [inline incubator/zookeeper/templates/PodDisruptionBudget.yaml]: Rename `component` label to `app`.

File: incubator/zookeeper/templates/PodDisruptionBudget.yaml [hunk 2]
```diff
@@ -1,20 +1,20 @@
...
 spec:
   selector:
     matchLabels:
-      component: "{{ printf "zk-%s" .Release.Name  | trunc 24 }}"
+      component: {{ template "name" . }}
```
Comment (unguiculus) [inline incubator/zookeeper/templates/PodDisruptionBudget.yaml]: Selector should have `app` and `release` labels.

File: incubator/zookeeper/templates/Service-headless.yaml [hunk 3]
```diff
@@ -15,4 +15,4 @@ spec:
     name: leader-election
   clusterIP: None
   selector:
-    component: "{{ printf "zk-%s" .Release.Name  | trunc 24 }}"
\ No newline at end of file
+    component: {{ template "name" . }}
```
Comment (unguiculus) [inline incubator/zookeeper/templates/Service-headless.yaml]: Selector should have `app` and `release` labels.

File: incubator/zookeeper/templates/StatefulSet.yaml [hunk 4]
```diff
@@ -1,21 +1,21 @@
...
+    component: {{ template "name" . }}
 spec:
-  serviceName: "{{ printf "zk-hsvc-%s" .Release.Name | trunc 24 }}"
+  serviceName: {{ template "fullname" . }}-headless
   replicas: {{.Values.Servers}}
   template:
     metadata:
       labels:
-        component: "{{ printf "zk-%s" .Release.Name  | trunc 24 }}"
+        component: {{ template "name" . }}
```
Comment (unguiculus) [inline incubator/zookeeper/templates/StatefulSet.yaml]: Add `app` and `release` labels.

File: incubator/zookeeper/templates/StatefulSet.yaml [hunk 5]
```diff
@@ -26,7 +26,7 @@ spec:
                   - key: "component"
                     operator: In
                     values:
-                    - "{{ printf "zk-%s" .Release.Name  | trunc 24 }}"
+                    - {{ template "name" . }}
```
Comment (unguiculus) [inline incubator/zookeeper/templates/StatefulSet.yaml]: Selector should have `app` and `release` labels.

File: incubator/zookeeper/templates/StatefulSet.yaml [hunk 6]
```diff
@@ -37,8 +37,8 @@ spec:
                     matchExpressions:
                       - key: "component"
                         operator: In
-                        values: 
-                         - "{{ printf "zk-%s" .Release.Name  | trunc 24 }}"
+                        values:
+                         - {{ template "name" . }}
```
Comment (unguiculus) [inline incubator/zookeeper/templates/StatefulSet.yaml]: Selector should have `app` and `release` labels.

**Review by unguiculus — CHANGES_REQUESTED** [review]
Thanks for this. Much appreciated.

**Comment by h0tbird** [comment]
Thank you @unguiculus I have added your requested changes.

**Comment by h0tbird** [comment]
Notify https://github.com/kubernetes/charts/issues/233 whenever this is merged.

File: incubator/zookeeper/templates/PodDisruptionBudget.yaml [hunk 7]
```diff
@@ -1,20 +1,21 @@
...
+  name: {{ template "fullname" . }}
   annotations:
     helm.sh/created: {{.Release.Time.Seconds | quote }}
   labels:
-    heritage: {{.Release.Service | quote }}
-    release: {{.Release.Name | quote }}
+    app: "{{ template "name" . }}"
     chart: "{{.Chart.Name}}-{{.Chart.Version}}"
-    component: "{{ printf "zk-%s" .Release.Name  | trunc 24 }}"
+    release: {{.Release.Name | quote }}
```
Comment (unguiculus) [inline incubator/zookeeper/templates/PodDisruptionBudget.yaml]: Please use consistent formatting. Add space after `{{` and before `}}`. Please fix this everywhere in the chart.

**Review by unguiculus — CHANGES_REQUESTED** [review]

**Comment by h0tbird** [comment]
@unguiculus now formatting is consistent.

**Comment by unguiculus** [comment]
Any reason why you are now using `include` vs. `template` for the `app` label? That's uncommon.

While you are at it, would you mind removing the `helm.sh/created` annotation. It does more harm than good, especially on statefulsets.

Moreover, I'd suggest to have file names all lowercase. That's how it is commonly done.

**Comment by h0tbird** [comment]
I use `include` because it works in combination with `| quote` whereas `template` doesn't.
I tend to use the `quote` function over literal `""` because I don't like nesting quotes.
Nonetheless I am ok reverting this change if you don't like it.

**Review by unguiculus — APPROVED** [review]

**Comment by unguiculus** [comment]
@k8s-bot ok to test

**Comment by unguiculus** [comment]
Just tested it on GKE. Only one pod comes up. The results from the test are actually wrong.
https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/charts/1317/pull-charts-e2e/2148/?log#log

**Comment by h0tbird** [comment]
I can't reproduce the error on my test GKE cluster:
```
[0] ~ >> kubectl version
Client Version: version.Info{Major:"1", Minor:"7", GitVersion:"v1.7.0", GitCommit:"d3ada0119e776222f11ec7945e6d860061339aad", GitTreeState:"clean", BuildDate:"2017-06-29T23:15:59Z", GoVersion:"go1.8.3", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"6", GitVersion:"v1.6.6", GitCommit:"7fa1c1756d8bc963f1a389f4a6937dc71f08ada2", GitTreeState:"clean", BuildDate:"2017-06-16T18:21:54Z", GoVersion:"go1.7.6", Compiler:"gc", Platform:"linux/amd64"}
```

```
[0] ~ >> helm version            
Client: &version.Version{SemVer:"v2.5.0", GitCommit:"012cb0ac1a1b2f888144ef5a67b8dab6c2d45be6", GitTreeState:"clean"}
Server: &version.Version{SemVer:"v2.5.0", GitCommit:"012cb0ac1a1b2f888144ef5a67b8dab6c2d45be6", GitTreeState:"clean"}
```

Maybe it is related to https://github.com/kubernetes/helm/issues/2223? Is IPv6 disabled in your helm client machine? I will downgrade helm to `2.4.2` and test again.

**Comment by h0tbird** [comment]
Still running helm `2.5.0` I can partially reproduce the issue like this:
```
[0] ~ >> sudo sysctl net.ipv6.conf.all.disable_ipv6=1
net.ipv6.conf.all.disable_ipv6 = 1
```
```
[0] ~/git/charts/incubator/zookeeper (rename_files|✚1)>> helm install --timeout 600 --name zookeeper --namespace zookeeper --set Cpu=0.1 --set Memory=1Gi --set Heap=1G --set Storage=5Gi .
E0714 11:00:25.190395    5117 portforward.go:212] Unable to create listener: Error listen tcp6 [::1]:40249: bind: cannot assign requested address
NAME:   zookeeper
LAST DEPLOYED: Fri Jul 14 11:00:25 2017
NAMESPACE: zookeeper
STATUS: DEPLOYED

RESOURCES:
==> v1beta1/StatefulSet
NAME                 DESIRED  CURRENT  AGE
zookeeper-zookeeper  3        0        1s

==> v1beta1/PodDisruptionBudget
NAME                 MIN-AVAILABLE  ALLOWED-DISRUPTIONS  AGE
zookeeper-zookeeper  2              0                    1s

==> v1/Service
NAME                          CLUSTER-IP    EXTERNAL-IP  PORT(S)            AGE
zookeeper-zookeeper-headless  None          <none>       2888/TCP,3888/TCP  1s
zookeeper-zookeeper           10.3.252.216  <none>       2181/TCP           1s


NOTES:
Thank you for installing ZooKeeper on your Kubernetes cluster. More information
about ZooKeeper can be found at https://zookeeper.apache.org/doc/current/

1. ZooKeeper is not accessible outside of the Kubernetes cluster. Its purpose is
to provide coordination for distributed systems running inside the cluster. As
ZooKeeper uses a TCP based protocol with an internal wire format, you will
probably want to use an existing client library for communication with the
ensemble.

2. The officially maintained clients are written in C and Java, and these can be
obtained with ZooKeeper release from
http://www-us.apache.org/dist/zookeeper/zookeeper-3.4.9/.
A list of language specific bindings and higher level libraries is available
here https://cwiki.apache.org/confluence/display/ZOOKEEPER/ZKClientBindings.

3. Most ZooKeeper clients require a connection string when instantiating an
instance of the client library, or when first connecting the client to the
ensemble. The connection string takes the form
<server>:<port>,<server>:<port>,... . If your ensemble has greater than 1
server, you should provide multiple servers in the connection string to allow
for the client to continue to function in the presence of individual server
failures. It is important that you do not use IP addresses and instead use the
DNS entries for the domain established by the StatefulSet's HeadlessService.
Your connection string should look like zookeeper-zookeeper-headless-0:2181,zookeeper-zookeeper-headless-1:2181,...
You can also use the client Service zookeeper-zookeeper:2181 to connect to an
available ZooKeeper server.
```

But I still can successfully deploy zookeeper anyway.

**Comment by unguiculus** [comment]
@h0tbird I'm aware of the IPv6 issue. We have that too. From my experience that doesn't do any harm.

I managed to run it successfully now. My cluster did not meet the resource requirements. The defaults the Zookeeper chart specifies are kind of high. We nowadays recommend not to specify any default resources at all if possible.

**Comment by h0tbird** [comment]
I will remove defaults on Monday, I also think those are too high.

**Pull request merged**